metis-indexing/src/main/java/eu/europeana/indexing/FullBeanPublisher.java

```diff
@@ -159,4 +256,226 @@ private void publishToSolr(RdfWrapper rdf, FullBeanImpl fullBean) throws Indexin
+    final String combinedQueryOr = Stream.of(queryForDatasetIds, queryForMatchingFields)
+        .filter(StringUtils::isNotBlank).collect(Collectors.joining(" OR "));
+
+    //Create query to restrict search on specific datasetId subsets
+    final List<String> datasetIds = new ArrayList<>();
+    datasetIds.add(datasetId);
+    if (!CollectionUtils.isEmpty(datasetIdsToRedirectFrom)) {
+      datasetIdsToRedirectFrom.stream().filter(StringUtils::isNotBlank).forEach(datasetIds::add);
+    }
+    final String datasetIdSubsets = generateQueryInDatasetSubsets(datasetIds);
+    final String finalQuery = String.format("(%s) AND (%s)", datasetIdSubsets, combinedQueryOr);
+
+    //Apply solr query and execute
+    final Map<String, String> queryParamMap = new HashMap<>();
+    queryParamMap.put("q", finalQuery);
+
+    //Prepare sub-query
+    Predicate<SolrDocument> predicate = document -> {
+      boolean doesDocumentMatchWithQuery = false;
+      for (Entry<String, List<String>> entry : queryMatchingFieldsAndFirstQueryGroup.getRight()
+          .entrySet()) {
+        doesDocumentMatchWithQuery = document.getFieldValues(entry.getKey()).stream()
+            .map(String.class::cast)
+            .anyMatch(entry.getValue()::contains);
+      }
+      return doesDocumentMatchWithQuery;
+    };
+    //Preprocess sub-query and replace documents based on the result
+    SolrDocumentList solrDocuments = getSolrDocuments(queryParamMap);
+    final SolrDocumentList subQueryResults = solrDocuments.stream().filter(predicate)
+        .collect(Collectors.toCollection(SolrDocumentList::new));
+    if (!CollectionUtils.isEmpty(subQueryResults)) {
+      solrDocuments = subQueryResults;
+    }
+    //Return all identifiers found
+    if (!CollectionUtils.isEmpty(solrDocuments)) {
+      return solrDocuments.stream()
+          .map(document -> (String) document.getFieldValue(EdmLabel.EUROPEANA_ID.toString()))
+          .collect(Collectors.toList());
+    }
+    return new ArrayList<>();
+  }
+
+  private Pair<String, Map<String, List<String>>> generateQueryForMatchingFields(
+      RdfWrapper rdfWrapper) {
+    //Collect all required information for heuristics
+    final List<String> identifiers = rdfWrapper.getProviderProxyIdentifiers().stream()
+        .map(Identifier::getString).collect(Collectors.toList());
+    final List<String> titles = rdfWrapper.getProviderProxyTitles().stream().map(Title::getString)
+        .collect(Collectors.toList());
+    final List<String> descriptions = rdfWrapper.getProviderProxyDescriptions().stream()
+        .map(description -> {
+          if (StringUtils.isNotBlank(description.getString())) {
+            return description.getString();
+          } else if (description.getResource() != null && StringUtils
+              .isNotBlank(description.getResource().getResource())) {
+            return description.getResource().getResource();
+          }
+          return null;
+        }).collect(Collectors.toList());
```

Comment:
Should you not filter for null? This is apparently a possible outcome, and it would screw up the following logic.

Reply:
Fixed